examples/cpp/benchmark/SubscriberApp.cpp

```diff
@@ -0,0 +1,375 @@
+    subscriber_->get_default_datareader_qos(reader_qos);
+    reader_qos.reliability().kind = config.reliability;
+    reader_qos.durability().kind = config.durability;
+    reader_ = subscriber_->create_datareader(topic_sub_, reader_qos, this, StatusMask::all());
+    if (reader_ == nullptr)
+    {
+        throw std::runtime_error("DataReader initialization failed");
+    }
+}
+
+SubscriberApp::~SubscriberApp()
+{
+    if (nullptr != participant_)
+    {
+        // Delete DDS entities contained within the DomainParticipant
+        participant_->delete_contained_entities();
+
+        // Delete DomainParticipant
+        DomainParticipantFactory::get_instance()->delete_participant(participant_);
+    }
+}
+
+void SubscriberApp::on_publication_matched(
+        DataWriter* /*writer*/,
+        const PublicationMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        matched_ = static_cast<int16_t>(info.current_count);
+        std::cout << "Publisher matched." << std::endl;
+        cv_.notify_one();
+    }
+    else if (info.current_count_change == -1)
+    {
+        matched_ = static_cast<int16_t>(info.current_count);
+        std::cout << "Publisher unmatched." << std::endl;
+        stop();
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for PublicationMatchedStatus current count change" << std::endl;
+    }
+}
+
+void SubscriberApp::on_subscription_matched(
+        DataReader* /*reader*/,
+        const SubscriptionMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        std::cout << "Subscriber matched." << std::endl;
+    }
+    else if (info.current_count_change == -1)
+    {
+        std::cout << "Subscriber unmatched." << std::endl;
+        stop();
+    }
+    else
+    {
+        std::cout << info.current_count_change
+                  << " is not a valid value for SubscriptionMatchedStatus current count change" << std::endl;
+    }
+}
+
+void SubscriberApp::on_data_available(
+        DataReader* reader)
+{
+    SampleInfo info;
+    switch (msg_size_)
+    {
+        case CLIParser::MsgSizeKind::NONE:
+            while ((!is_stopped()) && (RETCODE_OK == reader->take_next_sample(&benchmark_, &info)))
+            {
+                if ((info.instance_state == ALIVE_INSTANCE_STATE) && info.valid_data)
+                {
+                    std::cout << "Sample with index: '" <<
+                        benchmark_.index() << "' (0 Bytes) RECEIVED" << std::endl;
+                    benchmark_.index(benchmark_.index() + 1);
+                    while (matched_ == 0)
+                    {
+                        std::unique_lock<std::mutex> initial_lock(mutex_);
+                        auto check = cv_.wait_for(initial_lock, std::chrono::milliseconds(1), [&]()
+                                        {
+                                            return is_stopped();
+                                        });
```

Comment:
We typically recommend not to perform blockign operations in callbacks. I think it would make more sens moving these lines to the first part of the `run()` method.
Additionally, I think we could change the `matched_` condition to be greater of equal to 2 since we expect both of out endpoints to be matched with the publisher ones.

Reply:
The idea is to make sure the subscriber not to sent back the data until the on_publication_match happen. To solve the case of the subscriber receiving the data but not sending it properly.
I don't understand how this could be moved to `run()`.
Also, I understand the comment of change for matched_ to be greater or equal to 2 is for the publisher and not here, right?

Reply:
The first thing in the `run()` method would be:
```
{
        // Wait for the data endpoints discovery
        std::unique_lock<std::mutex> matched_lock(mutex_);
        cv_.wait(matched_lock, [&]()
                {
                    // at least one has been discovered
                    return ((matched_ >= 2) || is_stopped());
                });
}
```
Then, I would remove all `while (matched_ == 0){}` blocks from the `on_data_available()`. This will make the subscriber to first wait to be matched and simplify the 'on_data_available()`. We should until greater or equal than 2 since both subscriber endpoints must be matched.

Reply:
Whats more, we could avoid waiting for any `matched_` within `run()`, since the reliability and durability will do the late-joining this for us when specified

Reply:
As we discussed in private chat, I just delated the blocking operation on the  `on_data_available()`.